x64dbg 'blocks' file rename and write operations on the debuggee

### What is the matter?
x64dbg should always open file like this:
GENERIC_READ, FILE_SHARE_READ | **FILE_SHARE_WRITE | FILE_SHARE_DELETE**

#### Why FILE_SHARE_WRITE and  FILE_SHARE_DELETE when I just need READ access?

Even if you wanna to open the file for read-only access you should **[allow others](https://docs.microsoft.com/en-us/windows/win32/fileio/creating-and-opening-files)** also to write the file and to do so you have to specify for the sharing flags `FILE_SHARE_WRITE`. For renaming `FILE_SHARE_DELETE `is required.
So in the context of `CreateFile()` this means:
```
dwDesiredAccess=GENERIC_READ
dwShareMode=FILE_SHARE_READ | **FILE_SHARE_WRITE  | FILE_SHARE_DELETE**
```

Why that is important? Well let's do some example.
0. Just check if you can rename calc.exe to mycalc.exe. (Rename it back to calc.exe)
1. start calc.exe. Keep it open and try rename it to mycalc.exe. That should work
2. Attach x64dbg to calc.exe  try renaming it now. You get an error and can't do it.
3. Detach x64dbg from calc.exe  try renaming it now. Now it works again.
___

### Example

The nice thing about `FILE_SHARE_WRITE | FILE_SHARE_DELETE` is that the patcher will be able to save changes to the current running debuggee.

So for this change the 'Backup' algo of the patcher like this.
1. RENAME Target.exe -> Target.exe.bak
2. Copy Target.exe.bak -> Target.exe.
Now you have Target.exe and Target.exe.bak as intented. 
The difference is now that you will be able to open Target.exe for write access.
( Well write access for Target.exe.bak will fail however that is not we needed)
___

### Ideas for a SelfcheckCheater plugin

Btw the rename trick can also be used to cheat file/filename based self checks. 
Let's there be Target.exe and the unmodifed copy Target.org
When debugger stops at first system breakpoint it'll just swap Target.exe Target.org. When it quits I'll swap it back.

Or to be more resilienten and safe change that algo like this:
Before load check if there is a Target.swap. If that is the case debugger somehow crash or was unable to swap back the files.
So run 'onQuit' handler now that does a proper swapback.
When loaded the actual Target.exe gets renamed to Target.swap. 
On quit that Target.swap gets Target.exe and the Target.exe is named to Target.org

So maybe it's time for a  SelfcheckCheater plugin. Would it possible to do so with current plugin API?

If you're interested here is my template for minhook to make hooking things trivially easy:

```cpp
#include "ntdll/ntdll.h"
#include <Windows.h>
#include "debug.h"
#include "MinHook/MinHook.h"
#include "vid.h"

extern "C" __declspec(dllexport) void surprise() { }

struct Hook
{
	const wchar_t* pszModule;
	const char* pszProcName;
	PVOID pDetour;
	PVOID* ppOriginal;
};

// Can's magic macro for passing macrofn(ESCAPE(T<int, int>))
#define ESCAPE(...) __VA_ARGS__

#pragma section(".hooks$1",long,read)
#pragma section(".hooks$2",long,read)
#pragma section(".hooks$3",long,read)
#pragma comment(linker, "/merge:hooks=.rdata")

// Can's magic2
__declspec(allocate(".hooks$1")) const Hook hooks_begin;
__declspec(allocate(".hooks$3")) const Hook hooks_end;

// You likely forgot about WINAPI
// error C2373: 'hook_Function': redefinition; different type modifiers
#define HOOK(Dll, ReturnType, Function) \
	static decltype(&Function) original_ ## Function; \
	static decltype(Function) hook_ ## Function; \
	__declspec(allocate(".hooks$2")) const Hook dupa ## Function = { L ### Function, #Function, hook_ ## Function, (LPVOID*)&original_ ## Function }; \
	static ReturnType hook_ ## Function

HOOK(vid.dll, BOOL WINAPI, VidRegisterCpuidHandler)(
	__in PT_HANDLE Partition,
	__in VID_PROCESSOR_INDEX ProcessorIndex,
	__in VID_QUEUE_HANDLE MessageQueue,
	__in UINT32 CpuidFunction,
	__in PVOID UserContext,
	__out HANDLER_REF* HandlerRef
)
{
	dlog();
	return original_VidRegisterCpuidHandler(Partition, ProcessorIndex, MessageQueue, CpuidFunction, UserContext, HandlerRef);
}

#include <iterator>

BOOL WINAPI DllMain(
	_In_ HINSTANCE hinstDLL,
	_In_ DWORD     fdwReason,
	_In_ LPVOID    lpvReserved
	)
{
	dlog();
	if (fdwReason == DLL_PROCESS_ATTACH)
	{
		if (MH_Initialize() != MH_OK)
		{
			dlogp("MH_Initialize failed");
			return FALSE;
		}
		for (auto hook = std::next(&hooks_begin); hook != &hooks_end; ++hook)
		{
			if (!MH_CreateHookApi(hook->pszModule, hook->pszProcName, hook->pDetour, hook->ppOriginal))
			{
				dlogp("MH_CreateHook(%s) failed", hook->pszProcName);
				return FALSE;
			}
		}
		if (MH_EnableHook(MH_ALL_HOOKS) != MH_OK)
		{
			dlogp("MH_EnableHook failed");
			return FALSE;
		}
	}
	return TRUE;
}
```

--- a/src/dbg/module.cpp
+++ b/src/dbg/module.cpp
@@ -864,6 +864,10 @@ bool ModLoad(duint Base, duint Size, const char* FullPath, bool loadSymbols)
         // Load the physical module from disk
         if(StaticFileLoadW(wszFullPath.c_str(), UE_ACCESS_READ, false, &info.fileHandle, &info.loadedSize, &info.fileMap, &info.fileMapVA))
         {
+            // Fix an anti-debug trick, which opens exclusive access to the file
+            CloseHandle(info.fileHandle);
+            info.fileHandle = (HANDLE)1; // Set to non-zero for TitanEngine compatibility
+
             GetModuleInfo(info, info.fileMapVA);
 
             Size = GetPE32DataFromMappedFile(info.fileMapVA, 0, UE_SIZEOFIMAGE);

